haystack_experimental/components/agents/agent.py

```diff
@@ -0,0 +1,264 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any, Dict, List, Optional
+
+from haystack import component, default_from_dict, default_to_dict, Pipeline
+from haystack.components.generators.chat.openai import OpenAIChatGenerator
+from haystack.components.joiners import BranchJoiner
+from haystack.components.routers.conditional_router import ConditionalRouter
+from haystack.dataclasses import ChatMessage
+from haystack.lazy_imports import LazyImport
+from haystack.utils import Secret, deserialize_secrets_inplace
+
+from haystack_experimental.components.tools import ToolInvoker
+from haystack_experimental.tools import Tool
+from haystack_experimental.dataclasses.state import State, _schema_from_dict, _schema_to_dict, _validate_schema
+
+with LazyImport(message="Run 'pip install anthropic-haystack' to use Anthropic.") as anthropic_import:
+    from haystack_integrations.components.generators.anthropic.chat.chat_generator import (
+        AnthropicChatGenerator,
+    )
+
+_PROVIDER_GENERATOR_MAPPING = {
+    "openai": OpenAIChatGenerator,
+    "anthropic": AnthropicChatGenerator,
+}
+
+
+@component
+class Agent:
+    """
+    A Haystack component that implements a tool-using agent with provider-agnostic chat model support.
+
+    The component processes messages and executes tools until a exit_condition condition is met.
+    The exit_condition can be triggered either by a direct text response or by invoking a specific designated tool.
+
+    ### Usage example
+    ```python
+    from haystack.dataclasses import ChatMessage
+    from haystack.tools.tool import Tool
+
+    tools = [Tool(name="calculator", description="..."), Tool(name="search", description="...")]
+    input_types = {"search_depth": int}
+    output_types = {"total_tokens": int}
+
+    agent = Agent(
+        model="anthropic:claude-3",
+        generation_kwargs={"temperature": 0.7},
+        tools=tools,
+        exit_condition="search",
+        input_variables=input_types,
+        output_variables=output_types
+    )
+
+    # Run the agent
+    result = agent.run(
+        messages=[ChatMessage.from_user("Find information about Haystack")],
+        search_depth=2
+    )
+
+    assert "messages" in result  # Contains conversation history
+    assert "total_tokens" in result  # Contains tool execution outputs
+    ```
+    """
+
+    def __init__( # pylint: disable=too-many-positional-arguments
```

Comment:
Maybe not now, but we should look at how we could enable streaming for Agents. Will probably need a streaming_callback in the init and run methods like our ChatGenerators